LCM/scripts/python3/OMS_MetaConfigHelper.py

```diff
@@ -0,0 +1,259 @@
+
+if enable_flag == True and disable_flag == True:
+    exitWithError("Error: Cannot use both disable and enable options at the same time.")
+
+if enable_flag == True:
+    printVerboseMessage("Enable flag set to True, setting mof to enabled mode.")
+    if os.path.isfile(omshelper_disable_path):
+        os.remove(omshelper_disable_path)
+if disable_flag == True:
+    printVerboseMessage("Disable flag set to True, setting mof to disabled mode.")
+    disableFileHandle = open(omshelper_disable_path, "w")
+    try:
+        disableFileHandle.write("a")
+    finally:
+        disableFileHandle.close()
+
+    os.system("chown omsagent " + omshelper_disable_path)
+
+if not os.path.isfile(omshelper_disable_path):
+    # source the omsadmin conf file and get the key/value pairs
+    printVerboseMessage("OMS config path being read: " + os.path.realpath(conf_path))
+    keyvals = source_file(conf_path)
+
+    # Looking for DSC_ENDPOINT and AGENT_GUID
+
+    if "DSC_ENDPOINT" not in keyvals or "AGENT_GUID" not in keyvals:
+        exitWithError("Error: Unable to find needed key/value pairs in " + conf_path)
+
+    DSC_ENDPOINT = keyvals["DSC_ENDPOINT"].strip()
+
+    # Cut off everything after Nodes in the URL including Nodes, if Nodes is in URL (it should be)
+    Nodes_loc = DSC_ENDPOINT.find("/Nodes")
+    if Nodes_loc != -1:
+        DSC_ENDPOINT = DSC_ENDPOINT[:Nodes_loc]
+    AGENT_GUID = keyvals["AGENT_GUID"].strip()
+
+    # Write out agentID
+    f = open(agentid_path, "w")
+    try:
+        f.write(AGENT_GUID)
+    finally:
+        f.close()
+
+    os.system("chown omsagent " + agentid_path)
+
+    # Write metaconfig
+    metaConfig = generate_meta_mof(DSC_ENDPOINT)
+else:
+    metaConfig = generate_push_meta_mof()
+
+f = open(metamof_path, "w")
+try:
+    f.write(metaConfig)
+finally:
+    f.close()
+
+os.system("chown omsagent " + metamof_path)
+
+if os.geteuid() == 0:
+    commandToRun = "su - omsagent -c '/opt/microsoft/omsconfig/Scripts/python3/SetDscLocalConfigurationManager.py -configurationmof " + metamof_path + "'"
+else:
+    commandToRun = "/opt/microsoft/omsconfig/Scripts/python3/SetDscLocalConfigurationManager.py -configurationmof " + metamof_path
+
+# Apply the metaconfig using SetDscLocalConfiguration
+proc = subprocess.Popen(commandToRun, stdout=subprocess.PIPE, stderr=subprocess.PIPE, shell=True, close_fds=True)
+exit_code = proc.wait()
+
+stdout, stderr = proc.communicate()
+stdout = stdout.decode() if isinstance(stdout, bytes) else stdout
+printVerboseMessage("Output from3: " + commandToRun + ": " + str(stdout))
+
+set_metaconfig_success_string = ""
+if "omsconfig" in helperlib.DSC_SCRIPT_PATH:
+    set_metaconfig_success_string = "Operation SendMetaConfigurationApply completed successfully."
+else:
+    set_metaconfig_success_string = "ReturnValue=0"
+
+if ((exit_code == 0) and (stderr == '' or (sys.version_info >= (3, 0) and stderr.decode(encoding = 'UTF-8') == '')) and (set_metaconfig_success_string in str(stdout))):
```

Comment:
Would you consider decoding this stderr before checking further

Reply:
Not sure what you mean by this

Reply:
Here the stderr is of type bytes and this is being compared with empty string. I think the comparison of empty string to empty bytes will not give you the intended comparison. Am I right ?

Reply:
and I think this file will be called only in case of python3 environment right. I think this check of python version itself will not hold much importance.